### PR TITLE
WIP- vendor containers/{image,storage}

### PR DESCRIPTION
**Pull request by vrothberg**
Avoid redundantly pulling the same image (or layer).

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
Do not merge this PR, it's used to increase confidence for the changes to storage and image by letting it pass Buildah's CI.

**Comment by vrothberg** [comment]
This PR needs to be updated. We also need to merge https://github.com/containers/buildah/pull/1400 first to make it pass.

**Comment by rhatdan** [comment]
@vrothberg #1400 Is merged can you rework this one?

**Comment by vrothberg** [comment]
> @vrothberg #1400 Is merged can you rework this one?

Sure. Already rebased and pushed.

**Comment by rhatdan** [comment]
Can you remove [WIP]?
LGTM

**Comment by vrothberg** [comment]
> Can you remove [WIP]?

Not until the PRs over [image](https://github.com/containers/image/pull/607), [storage](https://github.com/containers/storage/pull/307) and [skopeo](https://github.com/containers/skopeo/pull/619) are merged. This PR is testing the new copy-detection mechanism.

I can open a dedicated PR to vendor in the latest image/storage.

**Comment by vrothberg** [comment]
All tests are hitting a timeout. I'll vendor the storage changes first to exclude some potential error sources from image.

**Comment by vrothberg** [comment]
@mtrmac @nalind @TomSweeneyRedHat the integration tests don't seem to fire off as the last logs are calling `tests/test_runner.sh` and do not show any log afterward. https://github.com/containers/libpod/pull/2763 over at libpod is passing now. I try to wrap my head around it I fail to reproduce.

**Comment by vrothberg** [comment]
It ultimately suggests a deadlock but I don't see how.

**Comment by vrothberg** [comment]
Good news before the weekend. No deadlock and tests are passing now. I took a shot a `travis_wait` which is famous for suppressing output when a timeout is being hit.

I don't yet fully understand WHY we hit the timeout but removing `travis_wait` let's us pass again. One really annoying thing about `travis_wait` is that it does not print *any* log at the time out which complicates debugging.

I've already checked with @TomSweeneyRedHat who wants a head nod from @nalind to make sure that removing `travis_wait` does not introduce any undesired side-effect.

File: pkg/blobcache/blobcache.go [hunk 1]
```diff
@@ -416,6 +416,18 @@ func (s *blobCacheDestination) HasThreadSafePutBlob() bool {
 	return s.destination.HasThreadSafePutBlob()
 }
 
+func (s *blobCacheDestination) SupportsBlobLocks() bool {
+	return false
```
Comment (mtrmac) [inline pkg/blobcache/blobcache.go]: (Note that the cache object can wrap a `storageImageDestination` in `Commit`. I don’t quite think that it matters — we are only actually creating some new layers, so there should not be any concurrent downloads of that.)

**Comment by vrothberg** [comment]
Closing in favour of Closing in favour of https://github.com/containers/image/pull/611.

**Pull request closed without merging**